src/Compilers/Core/Portable/PEWriter/MetadataWriter.PortablePdb.cs

```diff
@@ -978,13 +977,21 @@ private void EmbedMetadataReferenceInformation(CommonPEModuleBuilder module)
                 kind: _debugMetadataOpt.GetOrAddGuid(PortableCustomDebugInfoKinds.CompilationMetadataReferences),
                 value: _debugMetadataOpt.GetOrAddBlob(builder));
 
-            static PEReader GetReader(ISymbol symbol)
-                => symbol switch
+            static PEReader GetReader(PortableExecutableReference peReference)
+            {
+                switch (peReference.GetMetadata())
                 {
-                    IAssemblySymbol assemblySymbol => assemblySymbol.GetMetadata().GetAssembly().ManifestModule.PEReaderOpt,
-                    IModuleSymbol moduleSymbol => moduleSymbol.GetMetadata().Module.PEReaderOpt,
-                    _ => null
-                };
+                    case AssemblyMetadata assemblyMetadata:
+                        {
+                            var assembly = assemblyMetadata.GetAssembly();
```

Comment:
>assemblyMetadata.GetAssembly() [](start = 43, length = 30)

It looks like the previous code, using `assemblySymbol.GetMetadata().GetAssembly()`, would re-use the `AssemblyMetadata.Data` and corresponding `PEAssembly`. Is there overhead in creating a new `AssemblyMetadata.Data` and `PEAssembly` here?

Reply:
This is the part of the change I'm unsure about. Is there a more canonical / efficient way of grabbing the `PEReader` for the provided `MetadataReference`? The original tactic of grabbing the `IAssemblySymbol` doesn't work because we don't create those for `extern alias` reference. I couldn't see any other obvious route to go here. 

Did think about first trying to go through the reference manager as the old code did and then falling back to this trick. But again though it potentially adds this overhead.

Reply:
> Is there a more canonical / efficient way of grabbing the PEReader for the provided MetadataReference?

It looks like the reference manager has a `GetReferencedAssemblyAliases()` collection that returns pairs of `(IAssemblySymbolInternal, Immutable<string>)`. Would that help?

---
In reply to: [605779556](https://github.com/dotnet/roslyn/pull/52321#discussion_r605779556) [](ancestors = 605779556)

Reply:
Let me try that out.

Reply:
Perfect. That ended up working out nicely.